Design/Usage/AlgorithmUsageStatistics.md

```diff
@@ -0,0 +1,82 @@
+
+* type - Algorithm, Interface, Feature
+* name - Identifying name, for algorithms this would be Algorithm and version
+* time - The datetime of execution start
+* duration - (optional) the execution duration
+* details - (optional) Further details - for algorithms this will be a properties string, or possibly the python string of the Algorithm call.
+
+These would be sent in addition to these fields in common with the startup Usage details:
+
+* User ID hash
+* host ID hash
+
+###New Code
+
+####API::UsageService
+
+We would create a new service class called Usage Service which would be responsible for collating, and sending
+all usage data.  This would centralise all the logic covering Usage Reporting including:
+
+1. Detecting if repoting is enabled
+1. Registering the startup of Mantid
+1. Sending Startup usage reports, immediately, and every 24 hours thereafter
+1. Registering feature usage, and storing in a feature usage buffer
+1. Sending Feature usage reports on application exit, and when the feature usage buffer is above a size threshold.
+
+###### Implementation Notes:
+
+1. Use Poco::Timer to handle the timed aspects of the class
+  1. The 24 resend of startup data
+  1. Checking every n minutes if the feature buffer above the threshold for sending.
+1. Registering Feature usage must be fast, just create the record and return.
+1. Do not lock the feature usage buffer for any longer than is absolutely necessary.
+1. Use a queue, rather than a vector for the feature usage buffer.
+1. Create multiple overloads for registerFeatureUsage() to make is easy for other developers, one should be specialised for algorithms.
+1. Internet calls should use the InternetHelper.
+
+####Server side API
+
+Clearly the django website https://github.com/mantidproject/webapp will need to be extended to accept feature usage reports.
+
+Initially we do not plan to define reports for the feature usage data, they will be queried direct from the database using SQL.
+Once a good understanding of the usefull aspects of the data clarifies we will look to integrating some reports into the API.
+
+###To Existing Code
+
+####API::FrameworkManager
+
+Currently the code to send usage reports is in the FrameworkManager (FrameworkManagerImpl::SendStartupUsageInfo).  
+As described above this change would create a UsageService which would be responsible for collating, and sending
+all usage data.  As such I would propose to move the execution code from Framework Manager to the UsageService class, 
+and replace it with a single call to UsageService::Instance().registerStartup().
+This would also allow a single place to allow the additional logic to resend startup usage calls after 24 hours of constant running.
+Currently Long running instances cause the usage statistics of Mantid to underreport.
+
+###API::Algorithm
+
+We would add a call to UsageService::Instance().registerFeatureUsage(*this) within the Execute method of the Algorithm base class,
```

Comment:
The next version of this will have an adjustment, that the UsageService will actally be called UsageReporter and be owned and exposed by the Configservice.  This is only accessed by the algorithm base class, which has many many links to the rest of the framework.